From dfc83d5cf6cbfa9056c4892d5ad0c3d3ac667d4f Mon Sep 17 00:00:00 2001
From: vkarri <vkarri@contoso.com>
Date: Tue, 21 Oct 2025 01:23:08 +0000
Subject: [PATCH]  Use close range syscall over blindly looping over all FD

getdtablesize is deprecated in Kernel 6.12 and this is returning a really large number.
So, hsflowd takes a lot of time to close all file descriptors and is blocking the service restart of hsflowd
from sflowmgrd.

---
 src/Linux/hsflowd.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/Linux/hsflowd.c b/src/Linux/hsflowd.c
index a0e03cd..8ae9929 100644
--- a/src/Linux/hsflowd.c
+++ b/src/Linux/hsflowd.c
@@ -6,6 +6,7 @@
 extern "C" {
 #endif
 
+#include <sys/syscall.h>
 #include "hsflowd.h"
 #include "cpu_utils.h"
 #include "cJSON.h"
@@ -1922,9 +1923,12 @@ extern "C" {
 	  exit(EXIT_FAILURE);
 	}
 
-	// close all file descriptors
+	// close all file descriptors using close_range() (Kernel 5.9+)
 	int i;
-	for(i=getdtablesize(); i >= 0; --i) close(i);
+	if(close_range(0, ~0U, 0) < 0) {
+	  myLog(LOG_ERR,"close_range failed: %s", strerror(errno));
+	  exit(EXIT_FAILURE);
+	}
 	// create stdin/out/err
 	// stdin
 	if((i = open("/dev/null",O_RDWR)) == -1) {
-- 
2.39.5

